File: apps/proxy/internal/proxy/points.go
package proxy

import (
	"context"
	"errors"
	"fmt"
	"math"
	"time"

	"github.com/jackc/pgx/v5"

	appdb "github.com/opendum/opendum/apps/proxy/internal/db"
)

const (
	initialPointBalance  = 15
	roamingMinimumPoints = 1
	pointsPerMillion     = 1_000_000
)

type pointReservation struct {
	UserID  string
	Model   string
	Amount  int
	DebitID string
}

// roamingPoints converts token usage into points using the model's price,
// which is stored as points per million tokens. Successful roaming requests
// cost at least roamingMinimumPoints.
func (s *Service) roamingPoints(model string, usage *usageCounts) int {
	points := 0.0
	if usage != nil && s.registry != nil {
		if cost := s.registry.ModelCost(model); cost != nil {
			billableInput := usage.inputTokens - usage.cachedTokens
			if billableInput < 0 {
				billableInput = 0
			}
			points = (float64(billableInput)*cost.Input +
				float64(usage.cachedTokens)*cost.CacheRead +
				float64(usage.outputTokens)*cost.Output +
				float64(usage.cacheWriteTokens)*cost.CacheWrite) / pointsPerMillion
		}
	}
	total := int(math.Ceil(points))
	if total < roamingMinimumPoints {
		total = roamingMinimumPoints
	}
	return total
}

func (s *Service) reserveRoamingPoint(ctx context.Context, userID, model string) (*pointReservation, bool, error) {
	if userID == "" {
		return nil, true, nil
	}

	reservation := &pointReservation{
		UserID:  userID,
		Model:   model,
		Amount:  roamingMinimumPoints,
		DebitID: appdb.NewID(),
	}
	now := time.Now()

	tx, err := s.db.Pool.Begin(ctx)
	if err != nil {
		return nil, false, err
	}
	committed := false
	defer func() {
		if !committed {
			_ = tx.Rollback(ctx)
		}
	}()
	q := s.db.WithTx(tx)
	if err := ensurePointBalanceTx(ctx, q, userID, now); err != nil {
		return nil, false, err
	}

	balanceAfter, err := q.DebitPointBalance(ctx, appdb.DebitPointBalanceParams{
		Balance:   reservation.Amount,
		UpdatedAt: now,
		UserID:    userID,
	})
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, false, nil
		}
		return nil, false, err
	}

	err = q.InsertPointTransaction(ctx, appdb.InsertPointTransactionParams{
		ID:           reservation.DebitID,
		UserID:       userID,
		Amount:       -reservation.Amount,
		Type:         "roaming_debit",
		BalanceAfter: balanceAfter,
		CreatedAt:    now,
	})
	if err != nil {
		return nil, false, err
	}
	if err := tx.Commit(ctx); err != nil {
		return nil, false, err
	}
	committed = true

	return reservation, true, nil
}

func (s *Service) refundRoamingPoint(ctx context.Context, reservation *pointReservation) {
	if reservation == nil || reservation.UserID == "" || reservation.Amount <= 0 {
		return
	}

	now := time.Now()
	idempotencyKey := "roaming_refund:" + reservation.DebitID

	_ = func() error {
		tx, err := s.db.Pool.Begin(ctx)
		if err != nil {
			return err
		}
		committed := false
		defer func() {
			if !committed {
				_ = tx.Rollback(ctx)
			}
		}()
		q := s.db.WithTx(tx)
		if err := ensurePointBalanceTx(ctx, q, reservation.UserID, now); err != nil {
			return err
		}

		transactionID := appdb.NewID()
		affected, err := q.InsertPointTransactionOnConflictDoNothing(ctx, appdb.InsertPointTransactionOnConflictDoNothingParams{
			ID:             transactionID,
			UserID:         reservation.UserID,
			Amount:         reservation.Amount,
			Type:           "roaming_refund",
			BalanceAfter:   0,
			IdempotencyKey: &idempotencyKey,
			CreatedAt:      now,
		})
		if err != nil {
			return err
		}
		if affected == 0 {
			committed = true
			return tx.Commit(ctx)
		}

		balanceAfter, err := q.CreditPointBalance(ctx, appdb.CreditPointBalanceParams{
			Balance:   reservation.Amount,
			UpdatedAt: now,
			UserID:    reservation.UserID,
		})
		if err != nil {
			return err
		}

		if err := q.UpdatePointTransactionBalance(ctx, appdb.UpdatePointTransactionBalanceParams{
			BalanceAfter: balanceAfter,
			ID:           transactionID,
		}); err != nil {
			return err
		}
		committed = true
		return tx.Commit(ctx)
	}()
}

// adjustRoamingPoints settles the difference between the amount held when the
// roaming request started and the points the request actually cost. A positive
// delta debits the difference and may push the balance negative; a negative
// delta refunds it.
func (s *Service) adjustRoamingPoints(ctx context.Context, reservation *pointReservation, points int) {
	if reservation == nil || reservation.UserID == "" {
		return
	}
	delta := points - reservation.Amount
	if delta == 0 {
		return
	}

	now := time.Now()
	idempotencyKey := "roaming_settle:" + reservation.DebitID
	_ = func() error {
		tx, err := s.db.Pool.Begin(ctx)
		if err != nil {
			return err
		}
		committed := false
		defer func() {
			if !committed {
				_ = tx.Rollback(ctx)
			}
		}()
		q := s.db.WithTx(tx)
		if err := ensurePointBalanceTx(ctx, q, reservation.UserID, now); err != nil {
			return err
		}

		transactionID := appdb.NewID()
		affected, err := q.InsertPointTransactionOnConflictDoNothing(ctx, appdb.InsertPointTransactionOnConflictDoNothingParams{
			ID:             transactionID,
			UserID:         reservation.UserID,
			Amount:         -delta,
			Type:           "roaming_settle",
			BalanceAfter:   0,
			IdempotencyKey: &idempotencyKey,
			CreatedAt:      now,
		})
		if err != nil {
			return err
		}
		if affected == 0 {
			committed = true
			return tx.Commit(ctx)
		}

		var balanceAfter int
		if delta > 0 {
			balanceAfter, err = q.DebitPointBalanceAllowNegative(ctx, appdb.DebitPointBalanceAllowNegativeParams{
				Balance:   delta,
				UpdatedAt: now,
				UserID:    reservation.UserID,
			})
		} else {
			balanceAfter, err = q.CreditPointBalance(ctx, appdb.CreditPointBalanceParams{
				Balance:   -delta,
				UpdatedAt: now,
				UserID:    reservation.UserID,
			})
		}
		if err != nil {
			return err
		}

		if err := q.UpdatePointTransactionBalance(ctx, appdb.UpdatePointTransactionBalanceParams{
			BalanceAfter: balanceAfter,
			ID:           transactionID,
		}); err != nil {
			return err
		}
		committed = true
		return tx.Commit(ctx)
	}()
}

func (s *Service) settleRoamingPoint(ctx context.Context, ownerUserID string, reservation *pointReservation, model string, usage *usageCounts) {
	if reservation == nil {
		return
	}
	points := s.roamingPoints(model, usage)
	s.adjustRoamingPoints(ctx, reservation, points)
	s.creditSharingPoint(ctx, ownerUserID, reservation.DebitID, points)
}

func (s *Service) creditSharingPoint(ctx context.Context, ownerUserID, debitID string, amount int) {
	if ownerUserID == "" || debitID == "" || amount <= 0 {
		return
	}

	now := time.Now()
	idempotencyKey := "sharing_credit:" + debitID
	_ = func() error {
		tx, err := s.db.Pool.Begin(ctx)
		if err != nil {
			return err
		}
		committed := false
		defer func() {
			if !committed {
				_ = tx.Rollback(ctx)
			}
		}()
		q := s.db.WithTx(tx)
		if err := ensurePointBalanceTx(ctx, q, ownerUserID, now); err != nil {
			return err
		}

		transactionID := appdb.NewID()
		affected, err := q.InsertPointTransactionOnConflictDoNothing(ctx, appdb.InsertPointTransactionOnConflictDoNothingParams{
			ID:             transactionID,
			UserID:         ownerUserID,
			Amount:         amount,
			Type:           "sharing_credit",
			BalanceAfter:   0,
			IdempotencyKey: &idempotencyKey,
			CreatedAt:      now,
		})
		if err != nil {
			return err
		}
		if affected == 0 {
			committed = true
			return tx.Commit(ctx)
		}

		balanceAfter, err := q.CreditPointBalance(ctx, appdb.CreditPointBalanceParams{
			Balance:   amount,
			UpdatedAt: now,
			UserID:    ownerUserID,
		})
		if err != nil {
			return err
		}

		if err := q.UpdatePointTransactionBalance(ctx, appdb.UpdatePointTransactionBalanceParams{
			BalanceAfter: balanceAfter,
			ID:           transactionID,
		}); err != nil {
			return err
		}
		committed = true
		return tx.Commit(ctx)
	}()
}

func ensurePointBalanceTx(ctx context.Context, q *appdb.Queries, userID string, now time.Time) error {
	affected, err := q.InsertPointBalanceOnConflictDoNothing(ctx, appdb.InsertPointBalanceOnConflictDoNothingParams{
		UserID:    userID,
		Balance:   initialPointBalance,
		CreatedAt: now,
		UpdatedAt: now,
	})
	if err != nil {
		return err
	}

	if affected > 0 {
		idempotencyKey := fmt.Sprintf("initial:%s", userID)
		_, err = q.InsertPointTransactionOnConflictDoNothing(ctx, appdb.InsertPointTransactionOnConflictDoNothingParams{
			ID:             appdb.NewID(),
			UserID:         userID,
			Amount:         initialPointBalance,
			Type:           "initial_grant",
			BalanceAfter:   initialPointBalance,
			IdempotencyKey: &idempotencyKey,
			CreatedAt:      now,
		})
		return err
	}

	return nil
}

File: apps/proxy/src/core/points.ts
import { and, eq, sql } from "drizzle-orm";
import type { ProxyDB } from "../db/index.js";
import { schema } from "../db/index.js";
import { newID } from "../db/id.js";

export const initialPointBalance = 15;
export const roamingPointCost = 2;

export interface PointReservation {
  userId: string;
  amount: number;
  debitID: string;
}

export class InsufficientPointsError extends Error {
  constructor() {
    super("insufficient points");
    this.name = "InsufficientPointsError";
  }
}

async function ensurePointBalanceTx(db: ProxyDB, userId: string, now: Date): Promise<void> {
  const inserted = await db
    .insert(schema.userPointBalance)
    .values({ userId, balance: initialPointBalance, createdAt: now, updatedAt: now })
    .onConflictDoNothing({ target: schema.userPointBalance.userId });
  const rowsAffected = inserted.rowCount ?? 0;
  if (rowsAffected > 0) {
    const idempotencyKey = `initial:${userId}`;
    await db
      .insert(schema.pointTransaction)
      .values({ id: newID(), userId, amount: initialPointBalance, type: "initial_grant", balanceAfter: initialPointBalance, idempotencyKey, createdAt: now })
      .onConflictDoNothing({ target: schema.pointTransaction.idempotencyKey });
  }
}

export async function reserveRoamingPoint(db: ProxyDB | null, userId: string): Promise<[PointReservation | null, boolean, Error | null]> {
  if (userId === "") return [{ userId, amount: 0, debitID: "" }, true, null];
  if (!db) return [null, false, new Error("no db")];

  const reservation: PointReservation = { userId, amount: roamingPointCost, debitID: newID() };
  const now = new Date();

  try {
    await db.transaction(async (tx) => {
      await ensurePointBalanceTx(tx, userId, now);
      const result = await tx.execute(
        sql`UPDATE user_point_balance SET balance = balance - ${reservation.amount}, "updatedAt" = ${now} WHERE "userId" = ${userId} AND balance >= ${reservation.amount} RETURNING balance`,
      );
      const rows = result.rows as Array<{ balance: number }>;
      if (rows.length === 0) {
        throw new InsufficientPointsError();
      }
      const balanceAfter = Number(rows[0]!.balance);
      await tx.insert(schema.pointTransaction).values({
        id: reservation.debitID,
        userId,
        amount: -reservation.amount,
        type: "roaming_debit",
        balanceAfter,
        createdAt: now,
      });
    });
  } catch (error) {
    if (error instanceof InsufficientPointsError) return [null, false, null];
    return [null, false, error as Error];
  }

  return [reservation, true, null];
}

export async function refundRoamingPoint(db: ProxyDB | null, reservation: PointReservation | null): Promise<void> {
  if (!reservation || reservation.userId === "" || reservation.amount <= 0 || !db) return;

  const now = new Date();
  const idempotencyKey = "roaming_refund:" + reservation.debitID;

  try {
    await db.transaction(async (tx) => {
      await ensurePointBalanceTx(tx, reservation.userId, now);
      const transactionID = newID();
      const inserted = await tx
        .insert(schema.pointTransaction)
        .values({ id: transactionID, userId: reservation.userId, amount: reservation.amount, type: "roaming_refund", balanceAfter: 0, idempotencyKey, createdAt: now })
        .onConflictDoNothing({ target: schema.pointTransaction.idempotencyKey });
      if ((inserted.rowCount ?? 0) === 0) return;

      const result = await tx.execute(sql`UPDATE user_point_balance SET balance = balance + ${reservation.amount}, "updatedAt" = ${now} WHERE "userId" = ${reservation.userId} RETURNING balance`);
      const rows = result.rows as Array<{ balance: number }>;
      if (rows.length === 0) return;
      const balanceAfter = Number(rows[0]!.balance);
      await tx.update(schema.pointTransaction).set({ balanceAfter }).where(eq(schema.pointTransaction.id, transactionID));
    });
  } catch {
    // ignore refund failures
  }
}

export async function creditSharingPoint(db: ProxyDB | null, ownerUserID: string, debitID: string, amount: number): Promise<void> {
  if (ownerUserID === "" || debitID === "" || amount <= 0 || !db) return;

  const now = new Date();
  const idempotencyKey = "sharing_credit:" + debitID;

  try {
    await db.transaction(async (tx) => {
      await ensurePointBalanceTx(tx, ownerUserID, now);
      const transactionID = newID();
      const inserted = await tx
        .insert(schema.pointTransaction)
        .values({ id: transactionID, userId: ownerUserID, amount, type: "sharing_credit", balanceAfter: 0, idempotencyKey, createdAt: now })
        .onConflictDoNothing({ target: schema.pointTransaction.idempotencyKey });
      if ((inserted.rowCount ?? 0) === 0) return;

      const result = await tx.execute(sql`UPDATE user_point_balance SET balance = balance + ${amount}, "updatedAt" = ${now} WHERE "userId" = ${ownerUserID} RETURNING balance`);
      const rows = result.rows as Array<{ balance: number }>;
      if (rows.length === 0) return;
      const balanceAfter = Number(rows[0]!.balance);
      await tx.update(schema.pointTransaction).set({ balanceAfter }).where(eq(schema.pointTransaction.id, transactionID));
    });
  } catch {
    // ignore credit failures
  }
}
